### PR TITLE
Cross currency

### PR DESCRIPTION
**Pull request by FelixKleinpeter**

### CODE REVIEW

File: src/backlight/portfolio/portfolio.py [hunk 1]
```diff
@@ -90,14 +90,99 @@ def construct_portfolio(
...
+#         - mkt : list of marketdata for each asset
+#         - base_ccy : asset of reference
+
+#     """
+#     pl = pt.value()
+#     return pl.sum(axis=1)
+
+
+def homogenize_pl(
+    pt: Portfolio, mkt: List[MarketData], base_ccy: str = "USD"
```
Comment (keisuke-umezawa) [inline src/backlight/portfolio/portfolio.py]: For `base_ccy: str`, we need to define Currency type somewhere.

File: src/backlight/portfolio/portfolio.py [hunk 2]
```diff
@@ -90,14 +90,99 @@ def construct_portfolio(
...
+                     then an auto reference would be JPY in case of cross FX (EURUSD, USDJPY) we should convert USD pl from first asset
+                     to JPY using USDJPY market
+                      -> Job done a bit different : all assets are converted to base_ccy. The sum is still not done.    """
+    new_positions = []
+
+    # We compute the intersection of index between market datas and portfolio positions for later use
+    mkt_pos_intersection = mkt[0].index.intersection(pt._positions[0].index)
+    for position in pt._positions:
+        # ccy is the currency wich in which are expressed the element of the position, e.g. JPY for USDJPY
+        ccy = position.symbol[-3:]
```
Comment (keisuke-umezawa) [inline src/backlight/portfolio/portfolio.py]: If we add the info `unit` in `position`, we do not need to parse strings.

Comment (keisuke-umezawa) [reply src/backlight/portfolio/portfolio.py]: Although it is coding styles things, if we follow these things, the code become more readable automatically 😄 

- Avoid Deep Nesting e.g. at most, we can have 3 nests. c.f. https://code.tutsplus.com/tutorials/top-15-best-practices-for-writing-super-readable-code--net-8118
- Keep function length within less than 40 lines c.f. https://github.com/google/styleguide/blob/gh-pages/pyguide.md#318-function-length

Comment (FelixKleinpeter) [reply src/backlight/portfolio/portfolio.py]: Ok thanks. I will apply that.

File: src/backlight/portfolio/portfolio.py (hunk 1 above)
Comment (FelixKleinpeter) [reply src/backlight/portfolio/portfolio.py]: Should we create a new class in a new folder or add it in a file, like Portfolio.py?

Comment (keisuke-umezawa) [reply src/backlight/portfolio/portfolio.py]: How about creating `backlight/assets/currency.py`?

File: src/backlight/portfolio/portfolio.py [hunk 3]
```diff
@@ -90,14 +90,98 @@ def construct_portfolio(
     return Portfolio(positions)
 
 
-def calculate_pl(pt: Portfolio, mkt: List[MarketData]) -> pd.DataFrame:
+# def calculate_pl(pt: Portfolio, mkt: List[MarketData]) -> pd.DataFrame:
```
Comment (keisuke-umezawa) [inline src/backlight/portfolio/portfolio.py]: Delete comments.

File: src/backlight/portfolio/portfolio.py [hunk 4]
```diff
@@ -90,14 +90,98 @@ def construct_portfolio(
...
+#         - portfolio : a defined portfolio
+#         - mkt : list of marketdata for each asset
+#         - base_ccy : asset of reference
+
+#     """
+#     pl = pt.value()
+#     return pl.sum(axis=1)
+
+
+def homogenize_pl(
```
Comment (keisuke-umezawa) [inline src/backlight/portfolio/portfolio.py]: I think we do not need to have the `homogenize_pl`.

Please rename function as `calculate_pl`.

```
    pl = portforlio.value
    symbols = portforlio.symbols
    for symbol in symbols:
        if xxxx:
            pls.loc[:, symbol] = convert_positions(pls.loc[:, symbol])

    return pl
```

File: src/backlight/portfolio/portfolio.py [hunk 5]
```diff
@@ -90,14 +90,98 @@ def construct_portfolio(
...
+            new_positions.append(
+                position.loc[mkt[0].index.intersection(pt._positions[0].index)].copy()
+            )
+        else:
+            new_positions.append(convert_positions(position, mkt, ccy, base_ccy))
+
+    return Portfolio(new_positions)
+
+
+def convert_positions(
```
Comment (keisuke-umezawa) [inline src/backlight/portfolio/portfolio.py]: Make private function. Also, `convert_positions` is ambiguous.
How about naming it as `_convert_currency_unit`?

File: src/backlight/portfolio/portfolio.py (hunk 1 above)
Comment (keisuke-umezawa) [reply src/backlight/portfolio/portfolio.py]: `pt` -> `portfolio`.

File: src/backlight/portfolio/portfolio.py [hunk 6]
```diff
@@ -90,14 +90,98 @@ def construct_portfolio(
...
+#         - base_ccy : asset of reference
+
+#     """
+#     pl = pt.value()
+#     return pl.sum(axis=1)
+
+
+def homogenize_pl(
+    pt: Portfolio, mkt: List[MarketData], base_ccy: str = "USD"
+) -> Portfolio:
```
Comment (keisuke-umezawa) [inline src/backlight/portfolio/portfolio.py]: This function's result should be `pl`. So, `pd.Series`?

Comment (keisuke-umezawa) [reply src/backlight/portfolio/portfolio.py]: Please change implementation to return `pl`.

File: src/backlight/portfolio/portfolio.py [hunk 7]
```diff
@@ -90,14 +90,98 @@ def construct_portfolio(
...
+        - market : market forex datas
+        - ccy : the currency to convert from
+        - base_ccy : the currency to convert to
+    """
+    for market in mkt:
+        if ccy + base_ccy == market.symbol:
+            ratios = pd.Series(market.bid.values, index=market.index, dtype=float)
+        elif base_ccy + ccy == market.symbol:
+            ratios = pd.Series(market.ask.values, index=market.index, dtype=float)
+            ratios = ratios.apply(lambda x: 0 if x == 0 else 1.0 / float(x))
```
Comment (keisuke-umezawa) [inline src/backlight/portfolio/portfolio.py]: You do not need to use ask/bid here. Just using mid.

If you want to use ask/bid precisely here, we need to change ask/bid by current position size.

File: src/backlight/portfolio/portfolio.py [hunk 8]
```diff
@@ -90,14 +90,98 @@ def construct_portfolio(
...
+    new_p = Positions(
+        pd.DataFrame(
+            data=pos_values * ratios_values, columns=position.columns, index=idx
+        )
+    )
+    new_p.symbol = position.symbol
+    return new_p
+
+
+def get_ratios(mkt: List[MarketData], ccy: str, base_ccy: str) -> pd.Series:
```
Comment (keisuke-umezawa) [inline src/backlight/portfolio/portfolio.py]: Rename it as private function.

File: src/backlight/portfolio/portfolio.py (hunk 5 above)
Comment (keisuke-umezawa) [reply src/backlight/portfolio/portfolio.py]: If possible, I am very appreciate to add unittest.

File: src/backlight/portfolio/portfolio.py (hunk 8 above)
Comment (keisuke-umezawa) [reply src/backlight/portfolio/portfolio.py]: how about naming it as `_get_forex_ratios`?

File: src/backlight/portfolio/portfolio.py (hunk 4 above)
Comment (FelixKleinpeter) [reply src/backlight/portfolio/portfolio.py]: portfolio.value() function return the pl of each asset, while the convert_positions function takes a whole position. Do we need to apply the conversion+sum only to the pl of each asset? Thanks!

File: src/backlight/portfolio/portfolio.py [hunk 9]
```diff
@@ -90,14 +90,66 @@ def construct_portfolio(
...
+    pl: pd.Series, mkt: List[MarketData], ccy: str, base_ccy: str
+) -> pd.Series:
+    """
+    Convert the values of a position in a different currency from MarketData
+    args:
+        - position : the position to convert
+        - mkt : market forex datas
+        - ccy : the currency of the position
+        - base_ccy : the currency to express the position in
+    """
```
[the remark this answers is not in the document]
Comment (keisuke-umezawa) [reply src/backlight/portfolio/portfolio.py]: After that, you can write implementation more simple:

```
    idx = pl.index
    ratios = _get_forex_ratios(mkt[idx], ccy, base_ccy)
    return pl * ratios
```

File: tests/portfolio/test_portfolio.py [hunk 10]
```diff
@@ -128,3 +130,42 @@ def test_construct_portfolio(trades, markets, principal, lot_size):
...
+            columns=["amount", "price", "principal"],
+        )
+        p = Positions(df)
+        p.symbol = symbol
+        ptf.append(p)
+    return Portfolio(ptf)
+
+
+def test_calculate_pl(simple_portfolio, mid_markets):
+    calculated_portfolio = calculate_pl(simple_portfolio, mid_markets, base_ccy="usd")
```
Comment (keisuke-umezawa) [inline tests/portfolio/test_portfolio.py]: Could you add unittest for `base_ccy=jpy`?

**Comment by FelixKleinpeter** [comment]
The fix have been down. Should we merge?

File: src/backlight/portfolio/portfolio.py [hunk 11]
```diff
@@ -90,14 +90,66 @@ def construct_portfolio(
...
 
 
-def calculate_pl(pt: Portfolio, mkt: List[MarketData]) -> pd.DataFrame:
+def _convert_currency_unit(
+    pl: pd.Series, mkt: List[MarketData], ccy: str, base_ccy: str
+) -> pd.Series:
+    """
+    Convert the values of a position in a different currency from MarketData
+    args:
+        - position : the position to convert
```
[the remark this answers is not in the document]
Comment (keisuke-umezawa) [reply src/backlight/portfolio/portfolio.py]: `position : the position to convert` -> `pl`?

**Comment by FelixKleinpeter** [comment]
Sorry for my lack rigor. I fixed the comments.

**Pull request merged**